src/util/simplify_expr.cpp

```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
     return false;
   }
 
-  // eliminate casts to _Bool
-  if(expr_type.id()==ID_c_bool &&
-     op_type.id()!=ID_bool)
+  // circular casts through types shorter than `int`
+  const auto check_circular_cast=[&](const typet &type)
   {
-    // casts from boolean to a signed int and back:
-    // (boolean)(int)boolean -> boolean
-    if(expr.op0().id()==ID_typecast && op_type.id()==ID_signedbv)
+    if(expr_type==type)
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
+      if(typecast.op().type()==type)
       {
         expr=typecast.op0();
         return false;
       }
     }
+    return true;
+  };
+  if(op_type==signedbv_typet(32) &&
```

Comment:
Consider breaking this if up as it is hard to tell what the relation and reason of these conditions are